### PR TITLE
Task/cu 2r6ca2n/gravity gateway   fix loaded callback issue

### PR DESCRIPTION
**Pull request by thedev966**
Changes:
- Prevented multiple `loaded` callbacks issue

References:
https://app.clickup.com/t/2r6ca2n

### CODE REVIEW

File: src/master_gateway.js [hunk 1]
```diff
@@ -54,12 +54,18 @@ var masterGateway = {
...
         this.initialized = true;
         this.config = config;
-        this.slaves = slaves || {};
+        if (slaves) {
+            for (var slave in slaves) {
+                if (slaves.hasOwnProperty(slave)) {
+                    this.addSlave(slaves[slave]);
+                }
+            }
+        }
```
Comment (ZvonimirBusic) [inline src/master_gateway.js:68]: @thedev966 what is this doing, it doesn't seem like it is in task scope?

Comment (thedev966) [reply src/master_gateway.js:68]: @ZvonimirBusic Prior, there were two different ways to initialize slaves. So, Dinko and I concluded to stick with one. Regardless of whether `init() `or `addSlave()` directly are called, we should always go through `addSlave()` to register a slave.

Comment (ZvonimirBusic) [reply src/master_gateway.js:68]: @thedev966 Is this breaking change, if app has `slaves` to init, will it behave different than before this change?

**Review by dikaso — CHANGES_REQUESTED** [review]

File: src/master_gateway.js [hunk 2]
```diff
@@ -266,6 +268,7 @@ var masterGateway = {
 
     slaveLoaded : function(event, slaveData) {
         if (!slaveData.loaded || this.slaves[slaveData.frameId].isLoaded) {
+            logger.out('warn', '[GG] Master:', 'Slave is already loaded!', event.data);
```
Comment (dikaso) [inline src/master_gateway.js:271]: This could be now false positive if `loaded` callback is not provided.

**Review by dikaso — CHANGES_REQUESTED** [review]

File: src/master_gateway.js (hunk 2 above)
Comment (thedev966) [reply src/master_gateway.js:271]: @dikaso Okay, so shall I destructure this into if/else if block and add logger only in second case?

Comment (dikaso) [reply src/master_gateway.js:273]: Agree.

File: test/master_gateway.js [hunk 3]
```diff
@@ -77,6 +77,32 @@ describe('Master gateway instantiation', function() {
         assert.equal(instance.slaves.hasOwnProperty('dummy'), false);
     });
 
+    it('Should prevent another loaded callback if a slave is already loaded', function() {
+        instance.addSlave({
+            frameId : 'dummy',
+            slaveId : 'dummy',
+        });
+        instance.slaves['dummy'].isLoaded = true;
```
Comment (dikaso) [inline test/master_gateway.js:85]: You should call `instance.slaveLoaded({}, slaveData);` in this line too. This is cheating :D

**Review by dikaso — CHANGES_REQUESTED** [review]

File: test/master_gateway.js (hunk 3 above)
Comment (thedev966) [reply test/master_gateway.js]: @dikaso Nice catch :D

**Review by dikaso — APPROVED** [review]

**Review by MeSlav — APPROVED** [review]

**Comment by thedev966** [comment]
@ZvonimirBusic Can you please approve this PR, so I can proceed with QA?

**Review by ZvonimirBusic — APPROVED** [review]

**Pull request merged**